Models/PersistentShoppingCart.cs

```diff
@@ -0,0 +1,178 @@
+                attributeIdsToValues == null ||
+                !attributeIdsToValues.Any()) return true;
+
+            var product = _contentManager.Get(productId);
+            return _attributesDrivers.All(d => d.ValidateAttributes(product, attributeIdsToValues));
+        }
+
+        public void Add(int productId, int quantity = 1, IDictionary<int, ProductAttributeValueExtended> attributeIdsToValues = null) {
+            if (!ValidateAttributes(productId, attributeIdsToValues)) {
+                // If attributes don't validate, don't add the product, but notify
+                _notifier.Warning(T("Couldn't add this product because of invalid attributes. Please refresh the page and try again."));
+                return;
+            }
+
+            _persistentShoppingCartServices.AddItem(new ShoppingCartItem(productId, quantity, attributeIdsToValues));
+
+            _products = null;
+        }
+
+        public void AddRange(IEnumerable<ShoppingCartItem> items) {
+            foreach (var item in items) {
+                Add(item.ProductId, item.Quantity, item.AttributeIdsToValues);
+            }
+        }
+
+        public void Clear() {
+            _products = null;
+            _persistentShoppingCartServices.ClearCart();
+            UpdateItems();
+        }
+
+        public void Remove(int productId, IDictionary<int, ProductAttributeValueExtended> attributeIdsToValues = null) {
+            _persistentShoppingCartServices.RemoveItem(productId, attributeIdsToValues);
+            _products = null;
+        }
+
+        public void UpdateItems() {
+            _persistentShoppingCartServices.ConsolidateCart();
+            _products = null;
+        }
+
+        public IEnumerable<ShoppingCartQuantityProduct> GetProducts() {
+            if (_products != null) return _products;
+
+            var ids = Items.Select(x => x.ProductId);
+
+            var productParts =
+                _contentManager.GetMany<ProductPart>(ids, VersionOptions.Published,
+                new QueryHints().ExpandParts<TitlePart, ProductPart, AutoroutePart>()).ToArray();
+
+            var productPartIds = productParts.Select(p => p.Id);
+
+            var shoppingCartQuantities =
+                (from item in Items
+                 where productPartIds.Contains(item.ProductId) && item.Quantity > 0
+                 select new ShoppingCartQuantityProduct(item.Quantity, productParts.First(p => p.Id == item.ProductId), item.AttributeIdsToValues))
+                    .ToList();
+
+            return _products = shoppingCartQuantities
+                .Select(q => _priceService.GetDiscountedPrice(q, shoppingCartQuantities))
+                .Where(q => q.Quantity > 0)
+                .ToList();
+        }
+
+        public decimal Subtotal() {
+            return Math.Round(GetProducts().Sum(pq => Math.Round(pq.Price * pq.Quantity + pq.LinePriceAdjustment, 2)), 2);
+        }
+
+        public TaxAmount Taxes(decimal subTotal = 0) {
```

Comment:
I'm thinking this implementation should be moved elsewhere, as it would be identical in any implementation of a cart.